### PR TITLE
debug: add a new API Compact.

### PR DESCRIPTION
**Pull request by hicqu**
Add debug API Compact. The Compact is running in `Debugger.CpuPool`.
And this Compact is **by manual** and doesn't have metrics. Do we need to add metrics or share a same metrics with `CompactWorker`?

### CODE REVIEW

File: src/raftstore/store/debug.rs [hunk 1]
```diff
@@ -151,6 +152,14 @@ impl Debugger {
             Err(e) => Err(box_err!(e)),
         }
     }
+
+    pub fn compact(&self, cf: String, from: Vec<u8>, to: Vec<u8>) -> Result<()> {
+        try!(validate_db_and_cf(DB::KV, &cf));
+        let db = self.engines.kv_engine.clone();
+        CompactRunner::new(db)
```
Comment (siddontang) [inline src/raftstore/store/debug.rs]: bad style to use CompactRunner here. 

I suggest extracting a compact function in the util/rocksdb. Btw, we may compact raft RocksDB too.

File: src/raftstore/store/debug.rs [hunk 2]
```diff
@@ -153,12 +153,17 @@ impl Debugger {
...
-        CompactRunner::new(db)
-            .compact_range_cf(cf, Some(from), Some(to))
-            .map_err(|e| box_err!(e))
+    /// Compact the cf[start..end) in the db **by manual**.
+    pub fn compact(&self, db: DB, cf: &str, start: &[u8], end: &[u8]) -> Result<()> {
+        try!(validate_db_and_cf(db, cf));
+        let db = match db {
+            DB::KV => &self.engines.kv_engine,
+            DB::RAFT => &self.engines.raft_engine,
+            _ => unreachable!(),
```
Comment (siddontang) [inline src/raftstore/store/debug.rs]: don't panic in the debug API

File: src/raftstore/store/debug.rs [hunk 3]
```diff
@@ -151,6 +157,15 @@ impl Debugger {
             Err(e) => Err(box_err!(e)),
         }
     }
+
+    /// Compact the cf[start..end) in the db **by manual**.
+    pub fn compact(&self, db: DB, cf: &str, start: &[u8], end: &[u8]) -> Result<()> {
+        try!(validate_db_and_cf(db, cf));
+        let db = try!(self.get_db_from_type(db));
+        let handle = box_try!(get_cf_handle(db, cf));
+        compact_range(db, handle, Some(start), Some(end), true);
```
Comment (BusyJay) [inline src/raftstore/store/debug.rs]: Why true? /cc @zhangjinpeng1987

Comment (siddontang) [reply src/raftstore/store/debug.rs]: any update?

Comment (hicqu) [reply src/raftstore/store/debug.rs]: It's a mistake. I wil fix it.

**Comment by hicqu** [comment]
@zhangjinpeng1987 , PTAL, thanks.

File: src/raftstore/store/debug.rs [hunk 4]
```diff
@@ -50,13 +52,17 @@ impl Debugger {
         Debugger { engines }
     }
 
+    fn get_db_from_type(&self, db: DB) -> Result<&RocksDB> {
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/debug.rs]: Please use another name for type.

File: src/util/rocksdb/mod.rs [hunk 5]
```diff
@@ -348,6 +348,20 @@ pub fn roughly_cleanup_range(db: &DB, start_key: &[u8], end_key: &[u8]) -> Resul
...
+/// Compact the cf in the specified range by manual or not.
+pub fn compact_range(
+    db: &DB,
+    handle: &CFHandle,
+    start_key: Option<&[u8]>,
+    end_key: Option<&[u8]>,
+    exclusive_manual: bool,
+) {
+    let mut compact_opts = CompactOptions::new();
+    // `exclusive_manual == false` means manual compaction can concurrently run with this.
```
Comment (zhangjinpeng1987) [inline src/util/rocksdb/mod.rs]: Can concurrently run with other background compactions.

File: src/raftstore/store/debug.rs [hunk 6]
```diff
@@ -151,15 +157,24 @@ impl Debugger {
             Err(e) => Err(box_err!(e)),
         }
     }
+
+    /// Compact the cf[start..end) in the db **by manual**.
+    pub fn compact(&self, db: DBType, cf: &str, start: &[u8], end: &[u8]) -> Result<()> {
+        try!(validate_db_and_cf(db, cf));
+        let db = try!(self.get_db_from_type(db));
+        let handle = box_try!(get_cf_handle(db, cf));
+        compact_range(db, handle, Some(start), Some(end), false);
```
Comment (BusyJay) [inline src/raftstore/store/debug.rs]: What if user wants to compact the whole key ranges?

**Pull request merged**